### ISSUE TITLE
Optimizer fails to remove redundant branch

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#4578]
I am surprised by the optimized code produced by this program:

```racket
#lang racket/base
(#%plain-lambda (x1)
  (let* ([x2 (if (eq? x1 'special) 'special x1)]
         [x3 (if (eq? x2 'special) #f x2)])
    x3))
```

Running racket with `PLT_LINKLET_SHOW_CP0=1` yields essentially the following output:

```scheme
(lambda (x1_1.8)
  (let ([x2_2.9 (if (#2%eq? x1_1.8 'special) 'special x1_1.8)])
    (if (#2%eq? x2_2.9 'special) #f x2_2.9)))
```

I would expect the first `if` to be deleted.

I don’t know if missed optimizations rise to the level of being considered bugs, especially since Racket is very vague about what optimizations are guaranteed. But I think you can make the case that this is a “Macro Writer’s Bill of Rights” violation.

**Comment by lexi-lambda** [comment]
Here’s a simpler version of this example that I think illustrates the general principle more directly:

```racket
#lang racket/base
(#%plain-lambda (x y)
  (if (eq? x y) y x))
```

I would like this to be optimized to `(#%plain-lambda (x y) x)`, since I think that optimization is sound in general.

**Comment by gus-massa** [comment]
It's not a bug, it's a wish list addition request :) 

The first one should be working, so I'm very confused. Because cptype can "constant propagate" symbols form an `eq?` comparison. So it must work even in more general cases. So probably there is a bug. 

---

The second one is very difficult in the general case, for example

    (if (and (eq? x y) (eq? z w) (eq? x z)) y w)

The first problem is to keep the list of transitive equivalences in O(N) instead of O(N^2). According to my daughter and Wikipedia, it's possible, but I must check that the auxiliary data of the algorithm can be stored in the trees that are used inside cptypes.

Also, note that you can replace the last `y` by `w`, but not in the other direction, so only calculating the transitive equivalences is not enough.

Anyway, if the simple pattern with two variables appears in too many cases of idiomatic code, it is possible to add it as an special case. I prefer the cases that can be generalized, but in some occasions a special case is very useful. For example, `cp0` has a special case to detect the expansion of `or`, and in `BC` there is a special case for `(if (null? x) '() x)` that I never copied to CS but I probably should.

**Comment by gus-massa** [comment]
In the fist one, I made a mistake and I read

     (let* ([x2 (if (eq? x1 'special) x1 'special)]   ;==> A

instead of 

     (let* ([x2 (if (eq? x1 'special) 'special x1)]   ;==> B

`A` is working as intended.

`B` is hard in the general case, as your second example.

**Comment by eutro** [comment]
> The first problem is to keep the list of transitive equivalences in O(N) instead of O(N^2). According to my daughter and Wikipedia, it's possible, but I must check that the auxiliary data of the algorithm can be stored in the trees that are used inside cptypes.
Yes, using a union-find datastructure this should be possible. Union-ing an equivalence would be constant time, resolving which equivalence class an element belongs to would only amortized be constant time. Though of course you'll have to be careful if you backtrack and need to un-union (by storing the links in a separate map, perhaps, or the Prolog way of preserving a 'trail' of writes).

**Comment by lexi-lambda** [comment]
Let me give a little bit of context about where this actually comes up, which may help motivate a more specific optimization. The code in question appears in the expansion of a `lambda` with optional arguments, and the value is not `'special` but `unsafe-undefined`. This results in a redundant check in the expansion of the following code:

```racket
(lambda ([x unsafe-undefined])
  (if (eq? x unsafe-undefined) default-e x))
```

In particular, the expansion contains the following subexpression:

```racket
(#%plain-lambda (x1)
  (let ([x (if (eq? x1 unsafe-undefined) unsafe-undefined x1)])
    (if (eq? x unsafe-undefined) default-e x)))
```

This is quite silly. `lambda` could inspect each default value expression and omit the conditional when the corresponding expression is `unsafe-undefined`, but really I think it should not have to—hence why I think this can be considered a Macro Writer’s Bill of Rights violation.

I agree that solving the problem in sufficient generality to also handle `and` seems like it could be a lot of work. But I think this specific pattern of checking if a value is a single known special value is relatively common when implementing things like default arguments.

**Comment by mflatt** [comment]
That's the one example where I had seen the optimization opportunity, too. I wouldn't be surprised if it's the only way it shows up in our current code base; if so, maybe its best addressed by the keyword-argument expansion, avoiding the cost of checking when compiling any `if`. It's possible that the pattern appears somewhere else, though.

**Comment by lexi-lambda** [comment]
> That's the one example where I had seen the optimization opportunity, too. I wouldn't be surprised if it's the only way it shows up in our current code base

My initial reaction to this was a bit skeptical—in my mind, this pattern is functionally just a generalized version of `or` used whenever `#f` ought to be a legal value. That is, it’s certainly quite common to write

```racket
(define x (or x1 default-e))
```

which is of course equivalent to

```racket
(if (eq? x1 #f) default-e x1)
```

though no one would write it that way. Instead, they just write `(or x1 default-e)`, and that pattern is certainly recognized already. The expression

```racket
(lambda (x1 default)
  (let ([x2 (if x1 x1 #f)])
    (if x2 x2 default)))
```

is optimized to simply `(lambda (x1 default) (if x1 x1 default))`. (Mildly interestingly: the optimization still works if you replace the first condition with `(not x1)` and flip the branches, but it fails if you use `(eq? x1 #f)` instead.)

However, after some reflection, I think you are right that this pattern doesn’t actually show up very often with any value other than `#f`. The reason is that, while it is quite common to write code that checks for a unique “not provided” value using `eq?`, that value is (intentionally) unique to the definition that uses it. That is, it is common to write code like this:

```racket
(define not-given (gensym 'not-given))
(define (f [x not-given])
  (if (eq? x not-given)
      (use-default)
      (use-value x)))
```

If the code were to use `unsafe-undefined`, the optimization described in this issue would be applicable. But nobody actually writes code like that, at least in part because historically there was no designated value that the implementer of `f` could assume the caller would never provide. However, `unsafe-undefined` can technically serve this purpose now simply because its use by `lambda` means that passing it to `f` would be unsafe anyway.

Given that `unsafe-undefined` is uniquely privileged in this sense, it could in theory make sense to apply the optimization exclusively to uses of `unsafe-undefined`, perhaps even via a helper form `or/unsafe-undefined` that does the check. That said, I suspect the broader usage of `unsafe-undefined` is not something we really want to encourage, especially given this seems like pretty small potatoes as far as performance goes.

---

tl;dr: I am essentially convinced that this would not be terribly helpful in general.

Let me explain my specific use case, instead. I recently published the [case-kw-lambda](https://github.com/lexi-lambda/case-kw-lambda) package, and I have been trying to make it as good a citizen as possible with respect to both runtime and code size. While doing this, I noticed it currently generates a lot of redundant checks. I don’t see any way to avoid these checks because I have to expand to `lambda` (rather than `#%plain-lambda` or `case-lambda`) to produce a procedure that efficiently speaks the keyword protocol.

If `lambda` is altered to perform this static check, I can change `case-kw-lambda` to use `unsafe-undefined` as a default value, and that would resolve my immediate problem. I will have to change `case-kw-lambda` to perform the same static check so that macros that expand to `case-kw-lambda` will not have the same problem that I do, but that is probably fine, since this doesn’t seem likely to come up very often.
